tools/scanner/src/analysis.rs

```diff
@@ -0,0 +1,629 @@
+        }
+        self
+    }
+
+    pub fn is_supported(&self) -> bool {
+        (self.closures
+            + self.coroutines
+            + self.floats
+            + self.fn_defs
+            + self.fn_ptrs
+            + self.interior_muts
+            + self.raw_ptrs
+            + self.recursive_types
+            + self.mut_refs)
+            == 0
+    }
+}
+
+struct TypeVisitor<'a> {
+    metrics: &'a mut FnInputProps,
+    visited: HashSet<Ty>,
+}
+
+impl<'a> TypeVisitor<'a> {
+    pub fn visit_variants(&mut self, def: AdtDef, args: &GenericArgs) -> ControlFlow<()> {
+        for variant in def.variants_iter() {
+            for field in variant.fields() {
+                self.visit_ty(&field.ty_with_args(args))?
+            }
+        }
+        ControlFlow::Continue(())
+    }
+}
+
+impl<'a> Visitor for TypeVisitor<'a> {
+    type Break = ();
+
+    fn visit_ty(&mut self, ty: &Ty) -> ControlFlow<Self::Break> {
+        if self.visited.contains(ty) {
+            self.metrics.recursive_types += 1;
+            ControlFlow::Continue(())
+        } else {
+            self.visited.insert(*ty);
+            let kind = ty.kind();
+            match kind {
+                TyKind::Alias(..) => {}
+                TyKind::Param(_) => self.metrics.generics += 1,
+                TyKind::RigidTy(rigid) => match rigid {
+                    RigidTy::Coroutine(..) => self.metrics.coroutines += 1,
+                    RigidTy::Closure(..) => self.metrics.closures += 1,
+                    RigidTy::FnDef(..) => self.metrics.fn_defs += 1,
+                    RigidTy::FnPtr(..) => self.metrics.fn_ptrs += 1,
+                    RigidTy::Float(..) => self.metrics.floats += 1,
+                    RigidTy::RawPtr(..) => self.metrics.raw_ptrs += 1,
+                    RigidTy::Ref(_, _, Mutability::Mut) => self.metrics.mut_refs += 1,
+                    RigidTy::Adt(def, args) => match def.kind() {
+                        AdtKind::Union => self.metrics.unions += 1,
+                        _ => {
+                            let name = def.name();
+                            if def.is_box() {
+                                self.metrics.boxes += 1;
+                            } else if name.ends_with("UnsafeCell") {
+                                self.metrics.interior_muts += 1;
+                            } else {
+                                self.visit_variants(def, &args)?;
+                            }
+                        }
+                    },
+                    _ => {}
+                },
+                kind => unreachable!("Expected rigid type, but found: {kind:?}"),
+            }
+            ty.super_visit(self)
+        }
+    }
+}
+
+fn dump_csv<T: Serialize>(mut out_path: PathBuf, data: &[T]) {
+    out_path.set_extension("csv");
+    info(format!("Write file: {out_path:?}"));
+    let mut writer = WriterBuilder::new().delimiter(b';').from_path(&out_path).unwrap();
```

Comment:
Is there any specific reason to use delimiter `;` instead of comma `,`?

Reply:
Yes, because the name of the functions often contain `,` to separate generic parameters.